Automattic/wp-calypso
# Editor: Proofreading

Closes #306

This pull request seeks to enable a user to proofread their writing via the [AfterTheDeadline service](http://afterthedeadline.com).

![proofreading](https://cloud.githubusercontent.com/assets/1779930/13224226/8e861056-d955-11e5-874d-2a36e1b910d3.gif)

__Implementation notes:__

The plugin is largely a port of the [`atd.core.js`](https://github.com/Automattic/jetpack/blob/master/modules/after-the-deadline/atd.core.js) and [`editor_plugin.js`](https://github.com/Automattic/jetpack/blob/master/modules/after-the-deadline/tinymce/editor_plugin.js) modules, adapted for use in Calypso.

Main points of divergence include:
- Revised code styling to conform to Calypso lint rules
- Communicate with AfterTheDeadline via CORS, not proxy endpoint [[1]](https://github.com/Automattic/wp-calypso/blob/b6ccb4260905ca06494f9f6f26535a8360ed028f/client/components/tinymce/plugins/after-the-deadline/plugin.js#L167-L186)
- Persist "Always Ignore Suggestion" options to user preferences [[1]](https://github.com/Automattic/wp-calypso/blob/b6ccb4260905ca06494f9f6f26535a8360ed028f/client/components/tinymce/plugins/after-the-deadline/plugin.js#L188-L196) [[2]](https://github.com/Automattic/wp-calypso/blob/b6ccb4260905ca06494f9f6f26535a8360ed028f/client/components/tinymce/plugins/after-the-deadline/plugin.js#L403-L429)
- Use Calypso `lib/mixins/i18n` for string translation [[1]](https://github.com/Automattic/wp-calypso/blob/b6ccb4260905ca06494f9f6f26535a8360ed028f/client/components/tinymce/plugins/after-the-deadline/plugin.js#L207)
- Add proper default browser spellcheck disabling [[1]](https://github.com/Automattic/wp-calypso/blob/b6ccb4260905ca06494f9f6f26535a8360ed028f/client/components/tinymce/plugins/after-the-deadline/plugin.js#L431-L437)
- Use supported localized AfterTheDeadline service hostname if can be determined [[1]](https://github.com/Automattic/wp-calypso/blob/b6ccb4260905ca06494f9f6f26535a8360ed028f/client/components/tinymce/plugins/after-the-deadline/plugin.js#L150-L165)

Because this is a port of large preexisting files, it's recommended that you focus review on the items listed above. *It is not the intention at this time to convert existing files to ES6 standards.*

__Testing instructions:__

Verify that proofreading behaves as expected, particularly with regard to the items listed above in Implementation Notes.

1. Navigate to the [Calypso post editor](http://calypso.localhost:3000/post)
2. Select a site, if prompted
3. Note that a "Proofread Writing" button is included in the editor toolbar
4. Enter content in the editor content area, intentionally including a proofreading mistake
5. Note that your browser built-in spellcheck may flag the spelling mistake
6. Click the "Proofread Writing" button in the editor toolbar
7. Note that the spelling mistake is flagged with a solid red underline
8. Note that if browser built-in spellcheck was flagged in Step 5, it is no longer flagged
9. Click the spelling mistake
10. Note that the mistake reason is noted, including substitutions
11. Click a substitution, Ignore Suggestion, or Ignore Always
12. Note that...
 - If a substitution is clicked, the text is replaced
 - If Ignore Suggestion is clicked, the text is unflagged as being mistaken
 - If Ignore Always is clicked, the text is unflagged as being mistaken, and upon saving and reloading the page, the error is not flagged on subsequent Proofread Writing checks

__Open questions:__

- Should we add a feature flag, restricting this to particular environments pending further testing?
